src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/AvoidPotentiallyExpensiveCallWhenLogging.cs

```diff
@@ -0,0 +1,333 @@
+                if (SymbolEqualityComparer.Default.Equals(method.ConstructedFrom, _logMethod) ||
+                    method.ConstructedFrom.IsOverrideOrImplementationOfInterfaceMember(_logMethod))
+                {
+                    logLevelArgumentOperation = invocation.Arguments.GetArgumentForParameterAtIndex(0);
+
+                    return true;
+                }
+
+                // LoggerExtensions.Log and named variants (e.g. LoggerExtensions.LogInformation)
+                if (_logExtensionsMethodsAndLevel.TryGetValue(method, out logLevel))
+                {
+                    // LoggerExtensions.Log
+                    if (logLevel == LogLevelPassedAsParameter)
+                    {
+                        logLevelArgumentOperation = invocation.Arguments.GetArgumentForParameterAtIndex(invocation.IsExtensionMethodAndHasNoInstance() ? 1 : 0);
+                    }
+
+                    return true;
+                }
+
+                var loggerMessageAttribute = method.GetAttribute(_loggerMessageAttributeType);
+
+                if (loggerMessageAttribute is null)
+                {
+                    return false;
+                }
+
+                // Try to get the log level from the attribute arguments.
+                logLevel = loggerMessageAttribute.NamedArguments
+                    .FirstOrDefault(p => p.Key.Equals(Level, StringComparison.Ordinal))
+                    .Value.Value as int?
+                    ?? LogLevelPassedAsParameter;
+
+                if (logLevel == LogLevelPassedAsParameter)
+                {
+                    logLevelArgumentOperation = invocation.Arguments
+                        .FirstOrDefault(a => a.Value.Type?.Name.Equals(LogLevel, StringComparison.Ordinal) ?? false);
+
+                    if (logLevelArgumentOperation is null)
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            public bool IsGuardedByIsEnabled(IOperation currentOperation, IInvocationOperation logInvocation, int logLevel, IArgumentOperation? logLevelArgumentOperation)
+            {
+                var conditional = currentOperation.GetAncestor<IConditionalOperation>(OperationKind.Conditional);
+
+                // This is the base case where no ancestor conditional is found.
+                if (conditional is null)
+                {
+                    return false;
+                }
+
+                var conditionInvocations = conditional.Condition
+                    .DescendantsAndSelf()
+                    .OfType<IInvocationOperation>();
+
+                // Check each invocation in the condition to see if it is a valid guard invocation, i.e. same instance and same log level.
+                // This is not perfect (e.g. 'if (logger.IsEnabled(LogLevel.Debug) || true)' is treated as guarded), but should be good enough to prevent false positives.
```

Comment:
Yes, we should err on the side of false negatives; it's ok to not raise diagnostics when we ideally would, but it's not ok to raise diagnostics erroneously.